File: R/example3_5.R
# Example 3.5
# Testing for Structural Change

setwd("C:/Course17/ceR/R")
data<-read.table("http://web.pdx.edu/~crkl/ceR/data/cjx.txt",header=T,nrows=39)

year<-data$YEAR
X<-log(data$X)
L<-log(data$L1) 
K<-log(data$K1)

results<-lm(X ~ L+K)
summary(results)
results1<-lm(X~ L+K,subset=year<1949)
summary(results1)
results2<-lm(X~ L+K,subset=year>1948)
summary(results2)

## Calculate RSS and DF for restricted and unrestricted models
RSSr<-sum(results$residuals^2)
DFr<-results$df.residual
RSSur<-sum(results1$residuals^2)+sum(results2$residuals^2)
DFur<-results1$df.residual+results2$df.residual

## Computing the Chow test statistic (F-test)
Ftest<-((RSSr-RSSur)/(DFr-DFur))/(RSSur/DFur)
Ftest

## Calculate P-value
1-pf(Ftest, DFr-DFur, DFur)

rm(list=ls())

File: python/example3_5.py
"""
# Example 3.5
# Testing for Structural Change
"""

import numpy as np
import pandas as pd
from scipy import stats
import statsmodels.formula.api as smf

data=pd.read_csv('http://web.pdx.edu/~crkl/ceR/data/cjx.txt',sep='\s+',nrows=39)

year=data['YEAR']
X=np.log(data['X'])
L=np.log(data['L1']) 
K=np.log(data['K1'])

df = pd.concat([X, L, K], keys=['X', 'L', 'K'], axis=1) 
model = smf.ols(formula='X ~ L + K', data=df).fit()
print(model.summary())
model1 = smf.ols(formula='X ~ L + K', data=df, subset=year<1949).fit()
print(model1.summary())
model2 = smf.ols(formula='X ~ L + K', data=df, subset=year>1948).fit()
print(model2.summary())

# from statsmodels.iolib.summary2 import summary_col
# summary_col([model,model1,model2],stars=True,info_dict={"N":lambda x:(x.nobs)})

DFr = model.df_resid
RSSr = model.ssr
DF1 = model1.df_resid
RSS1 = model1.ssr
DF2 = model2.df_resid
RSS2 = model2.ssr
DFur = DF1+DF2
RSSur = RSS1+RSS2

# Chow test
Ftest = ((RSSr-RSSur)/(DFr-DFur))/(RSSur/DFur)
Ftest
# Calculate P-value
1-stats.f.cdf(Ftest, DFr-DFur, DFur)
# threshold = stats.f.ppf(0.95, DFr-DFur, DFur)

